src/server/pkg/obj/obj.go

```diff
@@ -222,8 +222,57 @@ type Client interface {
 	IsIgnorable(err error) bool
 }
 
+type checkedReadCloser struct {
+	io.ReadCloser
+	size  uint64
+	count uint64
+}
+
+func newCheckedReadCloser(size uint64, rc io.ReadCloser) io.ReadCloser {
+	return &checkedReadCloser{
+		ReadCloser: rc,
+		size:       size,
+	}
+}
+
+func (crc *checkedReadCloser) Read(p []byte) (int, error) {
+	count, err := crc.ReadCloser.Read(p)
+	crc.count += uint64(count)
+	if err != nil {
+		if errors.Is(err, io.EOF) && crc.count != crc.size {
+			return count, errors.Errorf("read stream ended after the wrong length, expected: %d, actual: %d", crc.size, crc.count)
+		}
+		return count, err
+	} else if crc.count > crc.size {
+		return count, errors.Errorf("read stream ended after the wrong length, expected: %d, actual: %d", crc.size, crc.count)
+	}
+	return count, nil
+}
+
+type wrapperClient struct {
```

Comment:
There is probably a more descriptive name for this. Maybe `checkedClient` or `validationClient`?